keps/sig-network/1880-multiple-service-cidrs/README.md

```diff
@@ -0,0 +1,954 @@
+reconcile loops that rebuild the state of the allocators based on the assigned IPs to the Services,
+this allows to support upgrades and skewed clusters.
+
+Clusters running with the new allocation model will not keep running the reconcile loops that
+keep the bitmap used by the allocators.
+
+Since the new allocation model will remove some of the limitations of the current model, skewed
+versions and downgrades can only work if the configurations are fully compatible, per example,
+current CIDRs are limited to a /112 max for IPv6, if an user configures a /64 to their IPv6 subnets
+in the new model, and IPs are assigned out of the first /112 block, the old allocator based in
+bitmap will not be able to use those IPs creating an inconsistency in the cluster.
+
+It will be required that those Services are recreated to get IP addresses inside the configured
+ranges, for consistency, but there should not be any functional problem in the cluster if the
+Service implementation (kube-proxy, ...) is able to handle those IPs.
+
+Example:
+
+- flags set to 10.0.0.0/20
+- upgrade to 1.25 with alpha gate
+- apiserver create a default ServiceCIDRConfig object for 10.0.0.0/20
+- user creates a new ServiceCIDR for 192.168.1.0/24
+- create a Service which gets 192.168.1.1
+- rollback or disable the gate
+- the apiserver repair loops will generate periodic events informing the user that the Service with the IP allocated
+  is not within the configured range
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: ServiceCIDRConfig
+  - Components depending on the feature gate: kube-apiserver, kube-controller-manager
+
+###### Does enabling the feature change any default behavior?
+
+The time to reuse an already allocated ClusterIP to a Service will be longer, since the ClusterIPs
+now depend on an IPAddress object that is protected by a finalizer.
+
+The bootstrap logic has been updated to deal with the problem of multiple apiservers with different
+configurations, making it more flexible and resilient.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+If the feature is disabled the old allocator logic will reconcile all the current allocated IPs
+based on the Services existing on the cluster.
+
+If there are IPAddresses allocated outside of the configured Service IP Ranges in the apiserver flags,
+the apiserver will generate events referencing the Services using IPs outside of the configured range.
+The user must delete and recreate these Services to obtain new IPs within the range.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+There are controllers that will reconcile the state based on the current created Services,
+restoring the cluster to a working state.
```

Comment:
If post disablement, services are recreated so all allocated IPs live in the original ranges, than it's clear all we need is the same logic as original enablement. However, if the sequence is:
 - Enable, create new ServiceCIDRConfig, create some new services in that new range.
 - Disable, but don't recreate offending services
 - Re-enable

Is there anything special we need to do?

Reply:
IIUIC, I think that nothing else is needed.
The controller will watch the Services, it will detect no IPAddress object associated exist, it will check that there is a ServiceCIDRConfig for the ClusterIP and allocate if ok, or warn that a Service exist with an IP that doesn't belong to a ServiceCIDRConfig

Reply:
we have a controller that reconciles assigned ClusterIPs to Services, that information is stored in etcd , with the Service and is the source of truth, we rebuild from that.
There can not be two services with the same IP, that will be a big bug in the allocator logic

Reply:
Right, but are these separate controllers? If the above sequence allows a moment to exist where there is no IPAddress, and there is a Service with that IP, unless the allocation logic looks at the actual services to verify non-duplication, then you have the potential for a problem. On the other hand, this may just result in a failure to create the new Service. The point is this - which of these two is true?

If the allocation logic ONLY checks for the IPAddress, then you cannot guarantee that there is not already a Service existing with the chosen IP, due to the sequence above. In that case, a new Service may get assigned an IP that an existing Service already has. Will that new Service then fail in any validation, preventing it from actually being stored? Or could it get written to storage with the duplicate IP?

If the allocation logic checks the IPAddress, and then double-checks that no Service exists (and has some sort of lock...which it probably wouldn't because of optimistic concurrency), then we should be OK.

Reply:
ic your point, the allocators has to rebuild their internal state from the Services too, check if this comment is related

https://github.com/kubernetes/enhancements/pull/3365/files#r902024376

 I think that this should be more explicit in the kep, allocators should rebuild their internal state from Services, and maybe wait until everything is in sync? an allocator should be able to do it if he watches Services, ServiceCIDRConfig and IPAddresses

Reply:
ok, it is here

> This allocator will have an informer watching Services, ServiceCIDRConfigs and IPAddresses, so it can have locally the
information needed to assign new IP addresses to Services.

Line517